casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala

```diff
@@ -128,10 +133,111 @@ object Proposer {
         maxSeqNum  = latestMsgs.find(_.sender == sender).map(_.senderSeq)
       } yield maxSeqNum.getOrElse(-1)
 
-    def createBlock(validatorIdentity: ValidatorIdentity) =
+    def createBlock(validatorIdentity: ValidatorIdentity): F[BlockCreatorResult] =
       for {
-        parentsPreState <- MultiParentCasper.getPreStateForNewBlock
-        result          <- BlockCreator.create(parentsPreState, validatorIdentity, shardId, dummyDeployOpt)
+        // merge pre state
+        preState <- MultiParentCasper.getPreStateForNewBlock
+        // misc
+        preStateHash      = preState.preStateHash
+        creatorsPk        = validatorIdentity.publicKey
+        creatorsId        = ByteString.copyFrom(creatorsPk.bytes)
+        creatorsLatestOpt = preState.justifications.find(_.sender == creatorsId)
+        nextSeqNum        = creatorsLatestOpt.map(_.seqNum + 1).getOrElse(0L)
+        nextBlockNum      = preState.justifications.map(_.blockNum).max + 1
+        parentHashes      = preState.justifications.map(_.blockHash)
+        finalBonds        = preState.fringeBondsMap
+        offenders         = preState.justifications.filter(_.validationFailed).map(_.sender)
+        // slashing
+        preStateBonds <- RuntimeManager[F].computeBonds(preStateHash.toByteString)
+        toSlash       = offenders intersect preStateBonds.filter { case (_, b) => b > 0 }.keySet
+        _             <- Log[F].info(s"Slashing senders: [${toSlash.map(_.show).mkString("; ")}]")
+        // epoch
+        changeEpoch = epochLength % nextBlockNum == 0
+        // attestation
+        // no need to attest if nothing meaningful to finalize.
+        dag <- BlockDagStorage[F].getRepresentation
+        conflictSet = {
+          val msgMap = dag.dagMessageState.msgMap
+          parentHashes.flatMap(msgMap(_).seen) -- preState.fringe.flatMap(msgMap(_).seen)
+        }
+        nothingToFinalize = conflictSet.toList
+          .traverse(BlockStore[F].getUnsafe)
+          .map(_.forall { b =>
+            b.state.systemDeploys.isEmpty && b.state.deploys.isEmpty
+          })
+        waitingForSupermajorityToAttest = {
+          val newlySeen = creatorsLatestOpt
+            .map { prev =>
+              prev.justifications.flatMap(dag.dagMessageState.msgMap(_).seen) --
+                parentHashes.flatMap(dag.dagMessageState.msgMap(_).seen)
+            }
+            .getOrElse(Set())
+          newlySeen.toList.traverse(BlockStore[F].getUnsafe).map { newBlocks =>
+            val newStateTransition =
+              newBlocks.exists(b => b.state.systemDeploys.nonEmpty || b.state.deploys.nonEmpty)
```

Comment:
Consider to introduce function `hasDeploys(block: BlockMessage)` and used here and in creation of `nothingToFinalize`.